x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/PackedValuesBlockHash.java

```diff
@@ -136,57 +139,220 @@ void add() {
-                    bytes.bytes()[nullByte] |= (byte) (1 << nullShift);
-                }
-            }
-            int ord = Math.toIntExact(hashOrdToGroup(bytesRefHash.add(bytes.get())));
-            ords.appendInt(ord);
+            fillBytesSv();
+            ords.appendInt(Math.toIntExact(hashOrdToGroup(bytesRefHash.add(bytes.get()))));
             addedValue(position);
         }
 
         private void addMultipleEntries() {
             ords.beginPositionEntry();
             int g = 0;
-            outer: for (;;) {
-                for (; g < groups.length; g++) {
-                    Group group = groups[g];
-                    group.bytesStart = bytes.length();
-                    if (group.encoder.read(group.valueOffset + group.loopedIndex, bytes) == 0) {
-                        assert group.valueCount == 1 : "null value in non-singleton list";
-                        int nullByte = g / 8;
-                        int nullShift = g % 8;
-                        bytes.bytes()[nullByte] |= (byte) (1 << nullShift);
-                    }
-                    ++group.loopedIndex;
-                }
+            do {
+                fillBytesMv(g);
+
                 // emit ords
-                int ord = Math.toIntExact(hashOrdToGroup(bytesRefHash.add(bytes.get())));
-                ords.appendInt(ord);
+                ords.appendInt(Math.toIntExact(hashOrdToGroup(bytesRefHash.add(bytes.get()))));
                 addedValueInMultivaluePosition(position);
 
-                // rewind
-                Group group = groups[--g];
-                bytes.setLength(group.bytesStart);
-                while (group.loopedIndex == group.valueCount) {
-                    group.loopedIndex = 0;
-                    if (g == 0) {
-                        break outer;
+                g = rewindKeys();
+            } while (g >= 0);
+            ords.endPositionEntry();
+            for (Group group : groups) {
+                group.valueOffset += group.valueCount;
+            }
+        }
+    }
+
+    @Override
+    public ReleasableIterator<IntBlock> lookup(Page page, ByteSizeValue targetBlockSize) {
+        return new LookupWork(page, targetBlockSize.getBytes(), DEFAULT_BATCH_SIZE);
+    }
+
+    class LookupWork implements ReleasableIterator<IntBlock> {
+        private final long targetBytesSize;
+        private final Page page;
+        private int position;
+
+        private IntBlock next;
+
+        LookupWork(Page page, long targetBytesSize, int batchSize) {
+            try {
+                this.page = page;
+                this.targetBytesSize = targetBytesSize;
+                initializeGroupsForPage(page, batchSize);
+                next = buildNext();
+            } finally {
+                if (next == null) {
+                    page.releaseBlocks();
+                }
+            }
+        }
+
+        @Override
+        public boolean hasNext() {
+            return next != null;
```

Comment:
I think it might make sense to flip these from "build early" to "build late" - but I've not figured out quite how to do that yet. We have the option either way.